### PR TITLE
[587] Migrate over the Find search results and related code

### PR DESCRIPTION
**Pull request by darokel**
### Context

https://trello.com/b/yOZLQ2IK/find-pub-sprint-board

### Changes proposed in this pull request

- Moves over the Find search results page and related code
- Fix related specs for migrated code to work in new codebase
- Organise all stylesheets so it's clear what's shared between both domains and which styles are domain specific

At the risk of this becoming even bigger, this PR's goal was to get the results page's styles and structure working so that it unblocks subsequent tickets. There are a still a few bits left to complete this page which will be done in subsequent tickets, specifically:

- Hook up the filters
- Hook up the actual search via the query params
- Decide which pagination style to go with (we have two different styles and two different gems which provide the functionality)

This PR is pretty much a carbon copy of the original Find code with some minor adjustments to make it work in Publish

### Guidance to review

Head to https://publish-teacher-training-pr-3076.london.cloudapps.digital/find/results and check it loads and looks like the Find counterpart. 

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by darokel** [comment]
[Course Search - Results page](https://trello.com/c/rUUocKgd/587-course-search-results-page)

**Review by defong — COMMENTED** [review]
> https://publish-teacher-training-pr-3076.london.cloudapps.digital/find/results

How broken is the pagination?

**Comment by darokel** [comment]
> https://publish-teacher-training-pr-3076.london.cloudapps.digital/find/results

It's not broken, we just have two different styles for pagination. The one currently shown is used in the support console. We've customised the Kaminari views based on the setup from Register. We've also customised pagy to have a different look. So it works find but just looks different to:

Screenshot 2022-10-26 at 09 29 53

We can decide later which one  to go with as we'll have to figure out a way render custom kaminari views for both the support console and find/pub views.

File: app/decorators/find/course_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,165 @@
+module Find
+  class CourseDecorator < ApplicationDecorator
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/decorators/find/course_decorator.rb]: agreed, done

**Comment by darokel** [comment]
> wrong-course [ wrong-course ](https://user-images.githubusercontent.com/470137/198054848-84c7e4e3-181f-44c4-bba5-963a6f734ed1.gif) [ ](https://user-images.githubusercontent.com/470137/198054848-84c7e4e3-181f-44c4-bba5-963a6f734ed1.gif)

This was hard coded from earlier work. Have connected it

File: .erb-lint.yml [hunk 2]
```diff
@@ -21,3 +21,5 @@ linters:
         Enabled: false
       Lint/UselessAssignment:
         Enabled: false
+      Naming/VariableNumber:
+        Enabled: false
```
Comment (defong) [inline .erb-lint.yml:25]: Do we have to relax another rule?

File: spec/components/find_interface/results/no_results_component_spec.rb [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            no_results_found?: true,
+          )
+          component = render_inline(described_class.new(results: results_view))
+
+          expect(component).to have_link("Learn more about teacher training in Scotland", href: "https://teachinscotland.scot/")
+        end
+      end
+
+      context "Wales" do
+        it "renders Scottish teacher training website" do
```
Comment (defong) [inline spec/components/find_interface/results/no_results_component_spec.rb:45]: Wales was never Scottish....

Comment (defong) [inline spec/components/find_interface/results/no_results_component_spec.rb:45]: Wales was never Scottish

**Review by defong — COMMENTED** [review]
Can we have previews for those components?

File: .erb-lint.yml (hunk 2 above)
Comment (darokel) [reply .erb-lint.yml:25]: Yes

File: app/assets/stylesheets/find_application.scss [hunk 4]
```diff
@@ -1,9 +1,26 @@
...
+@import "components/find/filter-layout";
+@import "components/find/filter";
+@import "components/find/filters";
+@import "components/find/map";
+
+@import "components/find/promoted-link";
+@import "components/find/search-results";
+@import "components/find/summary-card";
+@import "components/find/text-ellipsis";
+@import "components/find/toggle";
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/assets/stylesheets/find_application.scss]: done

**Comment by darokel** [comment]
> Can we have previews for those components?

Sure will do in a sep. ticket

File: app/components/find_interface/results/sort_by_component.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        options_for_select(results.sort_options, selected: params["sortby"].to_i || 0),
+        {},
+        {
+          class: "govuk-select",
+          onchange: "this.form.submit()",
+          role: "listbox",
+          data: { qa: "sort-form__options" },
+        },
+      ) %>
+      <%= form.submit("Update", name: nil, class: "govuk-button", data: { qa: "sort-form__submit" }) %>
```
Comment (defong) [inline app/components/find_interface/results/sort_by_component.html.erb:19]:
```suggestion
      <%= form.submit("Update", class: "govuk-button", data: { qa: "sort-form__submit" }) %>
```

File: app/components/find_interface/results/sort_by_component.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<div class="app-search-results-header">
+  <% if results.location_filter? %>
+    <p class="govuk-body">Sorted by distance</p>
+  <% else %>
+    <%= form_with(url: find_results_path, method: "get", skip_enforcing_utf8: true, class: "app-search-results-header__sort", data: { qa: "sort-form" }) do |form| %>
```
Comment (defong) [inline app/components/find_interface/results/sort_by_component.html.erb:5]:
```suggestion
    <%= form_with(url: find_results_path, class: "app-search-results-header__sort", data: { qa: "sort-form" }) do |form| %>
```

File: app/services/find/unescaped_query_string_service.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+module Find
+  class UnescapedQueryStringService
+    def initialize(base_path:, parameters:)
+      @base_path = base_path
+      @parameters = parameters
+    end
+
+    class << self
```
Comment (defong) [inline app/services/find/unescaped_query_string_service.rb:8]: Can we use https://github.com/DFE-Digital/teacher-training-api/blob/main/app/services/service_pattern.rb

File: app/view_objects/find/results_view.rb [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        .merge(subject_parameters)
+    end
+
+    def courses
+      @courses ||= RecruitmentCycle.current.courses.includes(
+        :enrichments,
+        subjects: [:financial_incentive],
+        site_statuses: [:site],
+        provider: %i[recruitment_cycle ucas_preferences],
+      ).findable.page(query_parameters[:page] || 1)
```
Comment (defong) [inline app/view_objects/find/results_view.rb:35]: we should be using https://github.com/DFE-Digital/teacher-training-api/blob/main/app/services/course_search_service.rb


```
Course.findable
# findable is a little weird due to the join it uses
```

File: app/view_objects/find/results_view.rb [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    def subject_parameters
+      query_parameters["subject_codes"].present? ? { "subject_codes" => query_parameters["subject_codes"].presence } : {}
+    end
+
+    def subject_codes
+      query_parameters["subject_codes"] || []
+    end
+
+    def all_subjects
+      @all_subjects ||= Subject.select(:subject_name, :subject_code).order(:subject_name).all
```
Comment (defong) [inline app/view_objects/find/results_view.rb:109]:
```suggestion
      @all_subjects ||= Subject.active.where.not(subject_code: nil).select(:subject_name, :subject_code).order(:subject_name).all
```

There is a few here that should not be... 

image


We have lost `humanities`!

File: app/view_objects/find/results_view.rb [hunk 10]
```diff
@@ -0,0 +1,260 @@
+module Find
+  # FIND:TODO Will be fleshed out in an upcoming ticket
+  class ResultsView
+    attr_reader :query_parameters
+
+    include ActionView::Helpers::NumberHelper
+
+    MAXIMUM_NUMBER_OF_SUBJECTS = 43
```
Comment (defong) [inline app/view_objects/find/results_view.rb:8]: Is this used?

File: app/views/find/result_filters/provider/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+          ) %>
+        </li>
+      <% end %>
+    </ul>
+
+    <hr class="govuk-section-break govuk-section-break--m govuk-section-break--visible">
+
+    <div>
+      <%= govuk_details(summary_text: "Try another provider") do %>
+        <%= form_with url: provider_path, method: :get, data: { "ga-event-form" => "Provider" } do |form| %>
```
Comment (defong) [inline app/views/find/result_filters/provider/new.html.erb:24]: `provider_path` looks wrong...

File: app/views/find/result_filters/provider/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+          <fieldset class="govuk-fieldset">
+            <div class="govuk-form-group">
+              <%= form.label :query, class: "govuk-label" do %>
+                <strong>School, university or other training provider</strong>
+                <div class="govuk-hint govuk-!-margin-bottom-0">Enter the name or provider code</div>
+              <% end %>
+              <%= form.text_field :query, class: "govuk-input" %>
+            </div>
+
+            <%= form.submit "Search again", name: nil, class: "govuk-button" %>
```
Comment (defong) [inline app/views/find/result_filters/provider/new.html.erb:34]:
```suggestion
            <%= form.submit "Search again", class: "govuk-button" %>
```

File: spec/components/find_interface/results/no_results_component_spec.rb (hunk 3 above)
Comment (defong) [reply spec/components/find_interface/results/no_results_component_spec.rb:45]:
```suggestion
        it "renders Welsh teacher training website" do
```

image

File: app/view_objects/find/results_view.rb (hunk 8 above)
Comment (darokel) [reply app/view_objects/find/results_view.rb:35]: As discussed, i'll leave that in the follow up tickets which covers this work. This is just to get the results page displaying

File: app/view_objects/find/results_view.rb (hunk 10 above)
Comment (darokel) [reply app/view_objects/find/results_view.rb:8]: It's used by another method in this class. Have only moved methods used directly in the results page at this point in time though.

File: app/services/find/unescaped_query_string_service.rb (hunk 7 above)
Comment (darokel) [reply app/services/find/unescaped_query_string_service.rb]: done

File: app/views/find/result_filters/provider/new.html.erb (hunk 11 above)
Comment (darokel) [reply app/views/find/result_filters/provider/new.html.erb]: Yup the root wasn't there, doesn't do anything at the moment so have just added it in to use the correct path

**Review by defong — APPROVED** [review]
Lets get this in so we can have a good Halloween.

**Pull request merged**